pkg/sql/schemachanger/scbuild/internal/scbuildstmts/dependencies.go

```diff
@@ -0,0 +1,272 @@
+
+	// ClusterSettings returns the current cluster settings, as in execCfg.
+	ClusterSettings() *cluster.Settings
+
+	// Statements returns the statements behind this schema change.
+	Statements() []string
+}
+
+// CatalogReader should implement descriptor resolution, namespace lookups, and
+// all such catalog read operations for the builder. The following contract must
+// apply:
+// - errors are panicked;
+// - caches are avoided at all times, we read straight from storage;
+// - MayResolve* methods return zero values if nothing could be found;
+// - MayResolve* methods ignore dropped or offline descriptors;
+// - MustReadDescriptor does not;
+// - MustReadDescriptor panics if the descriptor was not found.
+type CatalogReader interface {
+	tree.TypeReferenceResolver
+	tree.QualifiedNameResolver
+
+	// MayResolveDatabase looks up a database by name.
+	MayResolveDatabase(ctx context.Context, name tree.Name) catalog.DatabaseDescriptor
+
+	// MayResolveSchema looks up a schema by name.
+	MayResolveSchema(ctx context.Context, name tree.ObjectNamePrefix) (catalog.DatabaseDescriptor, catalog.SchemaDescriptor)
+
+	// MayResolveTable looks up a table by name.
+	MayResolveTable(ctx context.Context, name tree.UnresolvedObjectName) (catalog.ResolvedObjectPrefix, catalog.TableDescriptor)
+
+	// MayResolveType looks up a type by name.
+	MayResolveType(ctx context.Context, name tree.UnresolvedObjectName) (catalog.ResolvedObjectPrefix, catalog.TypeDescriptor)
+
+	// ReadObjectNamesAndIDs looks up the namespace entries for a schema.
+	ReadObjectNamesAndIDs(ctx context.Context, db catalog.DatabaseDescriptor, schema catalog.SchemaDescriptor) (tree.TableNames, descpb.IDs)
+
+	// MustReadDescriptor looks up a descriptor by ID.
+	MustReadDescriptor(ctx context.Context, id descpb.ID) catalog.Descriptor
+}
+
+// AuthorizationAccessor for checking authorization (e.g. desc privileges).
+type AuthorizationAccessor interface {
+
+	// CheckPrivilege verifies that the current user has `privilege` on
+	// `descriptor`.
+	CheckPrivilege(
+		ctx context.Context, descriptor catalog.Descriptor, privilege privilege.Kind,
+	) error
+
+	// HasAdminRole verifies if a user has an admin role.
+	HasAdminRole(ctx context.Context) (bool, error)
+
+	// HasOwnership returns true iff the role, or any role the role is a member
+	// of, has ownership privilege of the desc.
+	HasOwnership(ctx context.Context, descriptor catalog.Descriptor) (bool, error)
+}
+
+// BuilderState encapsulates the state of the planned schema changes, hiding
+// its internal state to anything that ends up using it and only allowing
+// state changes via the provided methods.
+type BuilderState interface {
+
+	// AddNode adds a node into the NodeAccumulator.
+	AddNode(dir scpb.Target_Direction, elem scpb.Element)
+
+	// ForEachNode iterates over the accumulated notes in the NodeAccumulator.
+	ForEachNode(fn func(status scpb.Status, dir scpb.Target_Direction, elem scpb.Element))
```

Comment:
do we like that this is exposed up to the `scbuildstmt` package? Seems better to remove it and leave the process of asking predicate questions of nodes to the context interface. We can defer that to a subsequent change.

Reply:
I'm ambivalent about this and I'm totally open to revisiting this. I would like to defer this conversation until after future work involving element decomposition. What we _really want_ is not predicates on the builder state but something more powerful, some kind of _view_ on what there is so far, in terms of elements.

Reply:
In other words this is going to change soon anyway so let's leave this as it is for the time being.

Reply:
Cool, we can leave it be for now. Regarding interfaces in terms of elements vs predicates, I only sort of agree. I agree that we want to decompose and that we want to use that decomposition to drive the implementation of the builder. Where I disagree is about how to expose that to the statement building business logic.

This is a bit pedantic, but I would not be mad about extremely high level functions which then all defer to a a data structure that represents a queryable view over the decomposition. I think such a technique will 1) aid in unit testing and code readability and 2) enforce a nice isolation between the complexity of processing statements and querying elements. 

tl;dr I agree with wanting to write exactly the code you're saying you want to write, but I want yet another layer of indirection to access it.

Reply:
What you describe is what I had in mind when I mentioned the need for a "view" earlier in this thread.